proposals/1406_multiple-idps/README.md

```diff
@@ -0,0 +1,493 @@
+unusable Pinniped installation.
+
+During an upgrade, an existing installation of the Supervisor would already have a FederationDomain CR defined without
+an "identityProviders" section. To enable smooth upgrades, the "identityProviders" section would be optional.
+
+- The Supervisor code already correctly handles the case when there are no identity provider CRs defined. No users can
+  log in using that FederationDomain.
+- To handle the case where there is exactly one identity provider CR defined, the controller could load that CR for use
+  in the FederationDomain. The "displayName" of the identity provider would be automatically configured to be the same
+  name as the CR. This allows old configurations to continue working after upgrade.
+- When there are multiple identity provider CRs defined, the controller can fail to load the FederationDomain and update
+  its status to include an error saying that using a FederationDomain when multiple identity provider CRs are created
+  requires using the "identityProviders" field on the FederationDomain. This handles the case where the
+  user adds multiple identity provider CRs after upgrading, but forgets to add the "identityProviders" field to the
+  FederationDomain.
+
+If an admin adds "identityProviders" to a pre-existing FederationDomain and changes the "displayName" of a pre-existing
+identity provider, then:
+1. Pre-existing user sessions would fail to refresh, causing those users to need to interactively log in again, since
+   the identity provider names and types are already stored in user sessions for use during refreshes. This code already
+   has sufficient protections to ensure that we can never accidentally use a different identity provider during refresh
+   compared to which was used during initial login, even if there is an accidental name collision (via UID comparisons).
+2. Pre-existing kubeconfigs would now refer to the wrong identity provider name, and would need to be regenerated.
+
+If an admin wants to add a pre-existing identityProvider to a pre-existing FederationDomain without interrupting
+pre-existing sessions or needing to generate new kubeconfigs, they could take care to make the "displayName" of
+the identity provider exactly match the name of the identity provider CR.
+
+#### Tests
+
+Lots of new unit and integration tests will be required for using multiple FederationDomains, multiple identity
+providers, and identity transformations and policies.
+
+#### New Dependencies
+
+https://github.com/google/cel-go would move from being an indirect dependency (via k8s libraries) to a direct dependency.
+
+#### Performance Considerations
+
+No problems are anticipated. CEL is up to the task from a performance point of view.
+
+#### Observability Considerations
+
+The status of FederationDomains will be updated to show new types of validation errors. Unexpected transformation errors
+during login attempts will be logged in the Pod logs.
+
+#### Security Considerations
```

Comment:
Should we embed any additional information into the token? Perhaps the `IDP ID` or `DisplayName` used during the login? Perhaps the unmodified `username`?

Note that the original `username` and `groups` can be configured by the supervisor admin to show up in the `additionalClaims` (at least for OIDC IDPS) so it's likely not worth directly embedding those in the claim.

Reply:
I thought about this some more. I guess that the downstream session would have all of the information, so that the Supervisor would have all the necessary data for itself. Original username and original groups are likely not useful for Supervisor as-is (because of the frequent and continuous upstream checking). So really there are two situations that we may want to consider:

* Token consumers may want to know which IDP the user used to log in. We could possibly make `DisplayName` available as an additional claim if admins so choose to map that property.
* Edge computing - would we need to keep the original information around for some reason? Do we even need to consider edge computing right now?

Reply:
Another thought: would additionalClaims be mapped from the upstream or the downstream (transformed) claims?

I think upstream claims make sense because downstream claims will already be put in the token. Right now the only transformed downstream claims are username and groups anyways.

Reply:
Adding the displayName of the IDP used to login into the ID token seems like it could be a good idea. The `sub` claim already tries to uniquely identify the user. Perhaps we should add the displayName of the IDP into the text of the `sub` claim when there are multiple IDPs configured.

Reply:
Using the additional claim mapping feature would map the untransformed claims directly from the upstream IDP's ID token (and maybe userinfo response). Identity transformations only apply to determining the `username` and `groups` claims in the downstream (Supervisor-issued) tokens. Identity transformations happen after the OIDC/LDAP/ActiveDirectoryIdentityProvider has finished its job of authenticating the user and determining their upstream username and group names. The additional claim mapping feature is only for OIDCIdentityProvider.

Reply:
Example subject claims from before the multiple IDP feature are:
- `ldaps://ldap.tools.svc.cluster.local?base=ou%3Dusers%2Cdc%3Dpinniped%2Cdc%3Ddev&sub=MTAwMA` for an LDAP IDP. The `base=` is the configured user search base. The `sub=` is the base64-encoded value of the configured UID attribute (which might be a binary value) for the logged in user.
- `https://dex.tools.svc.cluster.local/dex?sub=CiQwNjFkMjNkMS1mZTFlLTQ3NzctOWFlOS01OWNkMTJhYmVhYWESBWxvY2Fs` for an OIDC IDP. The `sub=` is the value of the `sub` claim from the upstream OIDC IDP's ID token.

In both of these cases, we could append something like `&idpName=my-idp-display-name` onto the end the existing `sub` value to make sure that it is unique for different configured IDP displayNames.

Reply:
I think we can handle the above concerns regarding the `sub` of the Supervisor-issued ID tokens in the implementation PR. Seems like too small of a detail to add to this proposal document.

Reply:
I like using `idpName` in the `sub` claim.